File: frontend/buddease/src/app/components/development/authenticateWithWix.ts
// authenticateWithWix.ts


// Function to authenticate with Wix API
async function makeAuthenticationRequest(): Promise<string> {
  const clientId = process.env.WIX_API_KEY;
  const clientSecret = process.env.WIX_API_SECRET;
  const authUrl = 'https://www.wix.com/oauth/access';

  if (!clientId || !clientSecret) {
    throw new Error('WIX_API_KEY or WIX_API_SECRET is not defined');
  }

  const response = await fetch(authUrl, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/x-www-form-urlencoded',
    },
    body: new URLSearchParams({
      grant_type: 'client_credentials',
      client_id: clientId,
      client_secret: clientSecret,
    }),
  });

  if (!response.ok) {
    throw new Error('Failed to authenticate with Wix API');
  }

  const data = await response.json();
  return data.access_token;
}
type PageData = {
  title: string;
  layout: string;
  components: Array<{
    type: string;
    data: any;
  }>;
};

// Function to authenticate with Wix API
async function authenticateWithWix() {
    // Make API request to obtain access token using authentication credentials
    const accessToken = await makeAuthenticationRequest(); // Implement this function
    return accessToken;
  }
  
  // Function to create a new web page on Wix
  async function createWebPage(pageData: PageData) {
    const accessToken = await authenticateWithWix();
  
    // Make API request to Wix API to create a new web page
    const response = await fetch('https://api.wix.com/v1/pages', {
      method: 'POST',
      headers: {
        Authorization: `Bearer ${accessToken}`,
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(pageData),
    });
  
    const createdPage = await response.json();
    return createdPage;
  }
  
  // Function to update web page content on Wix
  async function updateWebPageContent(pageId: string, updatedContent: Record<string, any>) {
    const accessToken = await authenticateWithWix();
  
    // Make API request to Wix API to update web page content
    const response = await fetch(`https://api.wix.com/v1/pages/${pageId}`, {
      method: 'PUT',
      headers: {
        Authorization: `Bearer ${accessToken}`,
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(updatedContent),
    });
  
    const updatedPage = await response.json();
    return updatedPage;
  }
  
  // Function to publish web page changes on Wix
  async function publishWebPage(pageId: string) {
    const accessToken = await authenticateWithWix();
  
    // Make API request to Wix API to publish web page changes
    const response = await fetch(`https://api.wix.com/v1/pages/${pageId}/publish`, {
      method: 'POST',
      headers: {
        Authorization: `Bearer ${accessToken}`,
      },
    });
  
    const publishedPage = await response.json();
    return publishedPage;
  }
  
  // Example usage
  const pageData = {
    title: 'New Page',
    layout: 'default',
    components: [], // Array of components to be added to the page
  };  
  createWebPage(pageData)
    .then((createdPage) => {
      console.log('New page created:', createdPage);
      // Update page content
      const updatedContent = {
        // Updated page content
      };
      return updateWebPageContent(createdPage.id, updatedContent);
    })
    .then((updatedPage) => {
      console.log('Page content updated:', updatedPage);
      // Publish changes
      return publishWebPage(updatedPage.id);
    })
    .then((publishedPage) => {
      console.log('Page published successfully:', publishedPage);
    })
    .catch((error) => {
      console.error('Error:', error);
    });
  













// // Example usage
// const pageData: PageData = {
//   title: 'New Page',
//   layout: 'default',
//   components: [],
// };

// createWebPage(pageData)
//   .then((createdPage) => {
//     console.log('New page created:', createdPage);
//     const updatedContent = {};
//     return updateWebPageContent(createdPage.id, updatedContent);
//   })
//   .then((updatedPage) => {
//     console.log('Page content updated:', updatedPage);
//     return publishWebPage(updatedPage.id);
//   })
//   .then((publishedPage) => {
//     console.log('Page published successfully:', publishedPage);
//   })
//   .catch((error) => {
//     console.error('Error:', error);
//   });

File: backend/database/authenticate_with_wix.py
import os
import requests
from dotenv import load_dotenv

# Load environment variables from .env file
load_dotenv()

def authenticate_with_wix(api_key, api_secret):
    """
    Authenticate with the Wix API using the provided API key and secret.
    Returns the access token if authentication is successful, None otherwise.
    """
    # Wix API authentication endpoint
    auth_url = "https://www.wix.com/oauth/access"
    
    # Request payload
    payload = {
        "grant_type": "client_credentials",
        "client_id": api_key,
        "client_secret": api_secret
    }
    
    try:
        # Send POST request to authenticate
        response = requests.post(auth_url, data=payload)
        if response.status_code == 200:
            # Authentication successful, extract access token from response
            access_token = response.json().get("access_token")
            return access_token
        else:
            # Authentication failed, print error message
            print("Failed to authenticate with Wix API. Status code:", response.status_code)
            return None
    except requests.exceptions.RequestException as e:
        # Handle request errors
        print("Error occurred during authentication:", e)
        return None

def fetch_user_data(access_token):
    """
    Fetch user data from the Wix API using the provided access token.
    Returns user data if successful, None otherwise.
    """
    # Wix API endpoint to fetch user data
    user_data_url = "https://www.wix.com/api/v1/user"
    
    try:
        # Send GET request to fetch user data
        headers = {"Authorization": f"Bearer {access_token}"}
        response = requests.get(user_data_url, headers=headers)
        if response.status_code == 200:
            # User data fetched successfully
            user_data = response.json()
            return user_data
        else:
            # Failed to fetch user data, print error message
            print("Failed to fetch user data from Wix API. Status code:", response.status_code)
            return None
    except requests.exceptions.RequestException as e:
        # Handle request errors
        print("Error occurred while fetching user data:", e)
        return None

# Example usage:
wix_api_key = os.getenv("WIX_API_KEY")
wix_api_secret = os.getenv("WIX_API_SECRET")

# Authenticate with Wix API
access_token = authenticate_with_wix(wix_api_key, wix_api_secret)
if access_token:
    # Authentication successful, fetch user data
    user_data = fetch_user_data(access_token)
    if user_data:
        # User data fetched successfully, do something with it
        print("User data:", user_data)
